Add pytorch style dataloader

dayo05 opened the pull request:
In pytorch there are DataLoader on torch.utils.data, Torchsharp has DataIterator on TorchSharp.Data.

In pytorch, Creating Dataset is easy.
```python
class CustomImageDataset(Dataset):
    def __init__(self, annotations_file, img_dir, transform=None, target_transform=None):
        self.img_labels = pd.read_csv(annotations_file)
        self.img_dir = img_dir
        self.transform = transform
        self.target_transform = target_transform

    def __len__(self):
        return len(self.img_labels)

    def __getitem__(self, idx):
        img_path = os.path.join(self.img_dir, self.img_labels.iloc[idx, 0])
        image = read_image(img_path)
        label = self.img_labels.iloc[idx, 1]
        if self.transform:
            image = self.transform(image)
        if self.target_transform:
            label = self.target_transform(label)
        return image, label
```
This is sample code that I found on pytorch official tutorial.
But in DataIterator, I don't know how to use that, It is not similar as pytorch style, I cannot found tutorial about it.

So I make dataloader using like this.
This style is very similar as what pytorch does.

This will make iterating data more easier.

nietras commented:
This is "hard-coded" to only two tensors which doesn't fit general case scenarios like what we have. In python this is handled by the dynamic type system and being able to handle tuples etc. This cannot be mirrored with a good experience in C# in my view. The best generalization is to base this on `Dictionary<string, Tensor>` (or `Dictionary<object, Tensor>`) if generalizing key.

Additionally, `Dataset` alone is not in itself useful, it needs to be designed together with `DataLoader`, which is then responsible for batching, randomization etc. And since we need to batch we need to be able to iterator over `Tensor`s and produce the same structured output, which is why `Dictionary<string, Tensor>` seems apt.

nietras commented on src/TorchSharp/Data/Dataset.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        public long Count();
+        public (Tensor, Tensor) GetTensor(object index);
+        public IEnumerable<object> GetDataEnumerable();
```
remove "public IEnumerable<object> GetDataEnumerable();" this should not be part of interface

nietras commented on src/TorchSharp/Data/Dataset.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public interface Dataset: IDisposable
+    {
+        public long Count();
```
should be property `int Count {get; }`

nietras commented on src/TorchSharp/Data/Dataset.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace TorchSharp.Data
+{
+    public interface Dataset: IDisposable
```
`IDataset` maybe?

dayo05 replied:
I did like that because pytorch does.

NiklasGustafsson replied:
IDataset is right for a .NET interface. 

The 'Dataset' class is abstract in PyTorch, so please change to an abstract class, or change the name.

nietras commented on src/TorchSharp/Data/Dataset.cs:
```diff
@@ -0,0 +1,14 @@
+    public interface Dataset: IDisposable
+    {
+        public long Count();
+        public (Tensor, Tensor) GetTensor(object index);
```
`Dictionary<string, Tensor> this[int index]`

nietras commented on src/TorchSharp/Data/DataLoader.cs:
```diff
@@ -0,0 +1,96 @@
+
+                if (shuffle)
+                {
+                    var r = new Random();
```
move to ctor and create seeded from ctor parameter

dayo05 replied:
I make private reset method and call that on ctor.

nietras commented on src/TorchSharp/Data/DataLoader.cs:
```diff
@@ -0,0 +1,96 @@
+                        .Select(x => new {Number = r.Next(), Item = x})
+                        .OrderBy(x => x.Number)
+                        .Select(x => x.Item)
+                        .GetEnumerator();
```
create an array of indexes from data set e.g. `Enumerable.Range(0, dataset.Count).ToArray()` then fischer yates shuffle that and keep this array of shuffled indexes as what will be iterated over, order by random is not a great way of doing shuffling. 

Fischer-Yates shuffle code is simple:
```csharp
       // Fisher-Yates shuffling
        internal static void Shuffle<T>(T[] values, Random random)
        {
            int n = values.Count;
            while (n > 1)
            {
                int k = random.Next(n);
                --n;
                T temp = values[n];
                values[n] = values[k];
                values[k] = temp;
            }
        }
```

nietras replied:
If not shuffling simply use indexes without shuffling, this means you can remove the GetDataEnumerable, it also fits better in potential multi-threading and pre-loading.

mfagerlund replied:
There are ways of visiting every index in "random" order, without enumerating the list up front;

https://lemire.me/blog/2017/09/18/visiting-all-values-in-an-array-exactly-once-in-random-order/
https://lemire.me/blog/2017/09/26/benchmarking-algorithms-to-visit-all-values-in-an-array-in-random-order/

Something like that would be better than the Fisher-Yates shuffle, especially if you're processing very large numbers of samples.

nietras commented on src/TorchSharp/Data/DataLoader.cs:
```diff
@@ -0,0 +1,96 @@
+
+            private Tensor dataTensor;
+            private Tensor labelTensor;
+            private (Tensor, Tensor) tmp;
```
remove these members, there are not necessary

dayo05 replied:
dataTensor, labelTensor is required for return current value.

dayo05 replied:
I change that to directly change Current value.

nietras reviewed (commented):
I've made some review comments too. :) I have more but just some quick remarks.

dayo05 commented:
TODO: Write test about it

NiklasGustafsson commented on src/TorchSharp/Data/Dataset.cs:
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Generic;
+using static TorchSharp.torch;
+
+namespace TorchSharp.Data
```
This should be in 'TorchSharp.torch.utils.data'

NiklasGustafsson commented:
This is such an important addition, and we must make sure to get it right. I would like to see some more extensive code examples on how to use this -- the user experience is essential.

dayo05 commented:
I'm making a new example for reflection about review

NiklasGustafsson commented:
I'm going to add a few things to the code comments in a review, but I have some questions:

1. This will always shuffle the data set. Are there situations when you do not want to shuffle?
2. Should we allow data loaders to accept custom shufflers? That is, if I think (I'd probably be wrong) that I have a better shuffle algorithm, shouldn't I be allowed to use that, instead?
3. There's no way to specify the device on which the 'Current' tensors end up.
4. Why does 'Current' return a dictionary? Is it input vs. labels?
5. How do I split a data set into train, test, and validation data subsets?
6. How do we add data augmentation support?

Also, it would be great to demonstrate how to use this by converting the Examples to use this instead of the ad hoc loading I added earlier.

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+        {
+            public static partial class data
+            {
+                public class DataLoader : IEnumerable<Dictionary<string, Tensor>>, IDisposable
```
Please add doc comments to all classes and (at least) all non-private members. You can copy information from the Pytorch docs, if there is any.

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+                    /// <summary>
+                    /// Create pytorch style dataloader
+                    /// </summary>
+                    /// <param name="dataset"></param>
```
Fill in the param elements.

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+                                if (isFinished())
+                                    break;
+                                dic = dataset.GetTensor(getNextValue());
+                                foreach (var x in currentKeys)
```
Wouldn't it be better to gather all the unsqueezed tensors in a list (or array) and then catenate them all together at the end. It would seem a lot more efficient, and will create fewer temporaries.

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+                            currentVal = 0;
+                        }
+
+                        public bool MoveNext()
```
This will create a lot of temporaries. Where are they disposed?

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+                            Current = dataset.GetTensor(getNextValue());
+                            var currentKeys = Current.Keys;
+                            foreach (var x in currentKeys)
+                                Current[x].unsqueeze_(0);
```
Couldn't this be done with a list or a tuple instead of a dictionary? What are the keys?

dayo05 replied:
Keys are defined by user, I use dictionary to get ambiguous amount of tensor.

NiklasGustafsson commented on src/TorchSharp/Dataset.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            public static partial class data
+            {
+                public abstract class Dataset : IDisposable
```
Doc comments, please.

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+using System;
```
Copyright notice, please.

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+        int offset;
+        int runningvalue;
+
+        public ShuffleGenerator(int size)
```
Please add a reference to where the algorithm comes from, even if it's just a Pytorch source file. A paper describing it would be best, but some attribution is important.

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+
+namespace TorchSharp.Utils
+{
+    public class ShuffleGenerator
```
Should this really be public? I'm assuming that it's not really intended to be used outside of TorchSharp, but I could be wrong.

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+            runningvalue = offset;
+        }
+
+        private int getCurrentValue()
```
This seems like Java-style naming. If this is a public Pytorch-based API, then use Python names. If it is not, then please use .NET naming conventions.

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+
+        static bool coprime(int u, int v) => gcd(u, v) == 1;
+
+        static int gcd(int u, int v)
```
Is this Greatest Common Divisor? I don't know much about which is the most efficient way to compute it, but this looks a lot more complex than some of the algorithms I found on stackoverflow:

https://stackoverflow.com/questions/18541832/c-sharp-find-the-greatest-common-divisor

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+            var min = size / 2;
+            maxrange = size;
+            prime = selectCoPrimeResev(min, size);
+            offset = new Random().Next(size);
```
We should probably allow a user to pass in a custom seed for Random().

NiklasGustafsson commented on src/TorchSharp/Utils/ShuffleGenerator.cs:
```diff
@@ -0,0 +1,89 @@
+            var min = size / 2;
+            maxrange = size;
+            prime = selectCoPrimeResev(min, size);
+            offset = new Random().Next(size);
```
Some of my torchvision code also relies on .NET Random() for RNG, but I'm wondering whether that is right. Maybe using Pytorch's RNG is the right thing? Not sure.

dayo05 replied:
I saw pytorch code and they are using numpy random, python random, std(in C++) random, etc. So I think .NET random will doesn't matter

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+                            this.shuffle = shuffle;
+                            reset();
+                        }
+
```
Java-style names don't make sense here.

NiklasGustafsson commented on src/TorchSharp/Dataset.cs:
```diff
@@ -0,0 +1,26 @@
+                    {
+                    }
+
+                    public abstract int Count { get; }
```
Since data sets can be enormous, Count and GetTensor (and anyone calling them) should probably use 'long' instead of 'int'. It's not often that you need more than 32-bit int for indices, but I think this is one of those rare situations.

The same goes for all other places in the code where the integer is an index into the data set.

NiklasGustafsson commented on src/TorchSharp/DataLoader.cs:
```diff
@@ -0,0 +1,114 @@
+                    {
+                        private Dataset dataset;
+                        private int batchSize;
+                        private Device device;
```
'device' is never used.

NiklasGustafsson commented:
@dayo05 -- thanks for all the work so far. I'll be taking some time off in December, but I'm looking forward to integrating this when I come back.

Thanks,

Niklas

dayo05 commented:
I'm busy now, I'll going to work on two weeks later :(

dayo05 commented:
@NiklasGustafsson I finished all my job, review please.

dayo05 commented:
I found shuffler is not correctly working. I'll fix ASAP.

dayo05 commented:
This is example for classifying fruits360 dataset
```cs
public class Fruits360: Dataset
{
    private List<string> Labels = new();
    private List<string> images = new();
    public Fruits360(bool isTrain, Device device)
    {
        var root = "/home/dayo/datasets/fruits-360/" + (isTrain ? "Training" : "Test");
        //Labels.AddRange(Directory.GetDirectories(root));
        foreach(var x in Directory.GetDirectories(root))
            images.AddRange(Directory.GetFiles(x));
        Labels.AddRange(Directory.GetDirectories(root).Select(x => x.Split('/')[^1]));
    }

    public override long Count => images.Count;

    public override Dictionary<string, torch.Tensor> GetTensor(long index)
    {
        var image = Image.Load<Rgb24>(images[(int) index], new JpegDecoder());
        using var r = tensor(image.GetPixelMemoryGroup()[0].Span.ToArray().Select(x => x.R / 255.0f).ToList(),
            new long[] {1, 100, 100});
        using var g = tensor(image.GetPixelMemoryGroup()[0].Span.ToArray().Select(x => x.G / 255.0f).ToList(),
            new long[] {1, 100, 100});
        using var b = tensor(image.GetPixelMemoryGroup()[0].Span.ToArray().Select(x => x.B / 255.0f).ToList(),
            new long[] {1, 100, 100});
        return new()
        {
            {"image", cat(new List<Tensor> {r, g, b}, 0)},
            {"label", tensor(Labels.IndexOf(images[(int)index].Split('/')[^2]), ScalarType.Int64)}
        };
    }
}
```
```cs
using var trainDataset = new Fruits360(true, CUDA);
using var testDataset = new Fruits360(false, CUDA);
using var train = new DataLoader(trainDataset, 256, true, CUDA);
using var test = new DataLoader(testDataset, 512, false, CUDA);

var model = new Fruits360Model(CUDA);
var optimizer = optim.Adam(model.parameters(), learningRate: 0.01);

foreach (var epoch in Range(1, 1000))
{
    model.Train();
    var batchId = 1;
    Console.WriteLine($"Epoch{epoch} running");
    foreach (var x in train)
    {
        optimizer.zero_grad();

        var prediction = model.forward(x["image"]);
        var output = functional.nll_loss(reduction: Reduction.Mean)(prediction, x["label"]);
        
        output.backward();
        optimizer.step();
        
        Console.Write($"\rTrain: epoch {epoch} {batchId * 1.0 / train.Count:P2} [{batchId} / {train.Count}] Loss: {output.ToSingle():F9}");
        batchId++;
        
        prediction.Dispose();
        output.Dispose();
        GC.Collect();
    }

    using (no_grad())
    {
        model.Eval();
        
        var testLoss = 0.0;
        var correct = 0;
        var total = 0L;
        var idx = 0;
        foreach (var x in test)
        {
            idx++;
            Console.Write($"\rTest running: {idx * 1.0 / test.Count:P2}");
            var prediction = model.forward(x["image"]);
            var output = functional.nll_loss(reduction: Reduction.Sum)(prediction, x["label"]);
            testLoss += output.ToSingle();

            var pred = prediction.argmax(1);
            total += pred.size()[0];
            correct += pred.eq(x["label"]).sum().ToInt32();
            pred.Dispose();
            prediction.Dispose();
            output.Dispose();
            GC.Collect();
        }
        Console.WriteLine(
            $"\rTest set: Average loss {(testLoss / testDataset.Count):F9} | Accuracy {((double) correct / testDataset.Count):P2}");
    }
}


class Fruits360Model : Module
{
    private Module layer1 = Sequential(
        Conv2d(3, 32, 3),
        ReLU(),
        MaxPool2d(2, 2));

    private Module layer2 = Sequential(
        Conv2d(32, 64, 3),
        ReLU(),
        MaxPool2d(2, 2));

    private Module layer3 = Sequential(
        Conv2d(64, 64, 3),
        ReLU(),
        MaxPool2d(2, 2));

    private Module fc = Sequential(
        Flatten(),
        Linear(6400, 1024),
        ReLU(),
        Dropout(),
        Linear(1024, 625),
        ReLU(),
        Linear(625, 131));
    public Fruits360Model(Device? device) : base("fruits360")
    {
        RegisterComponents();
        to(device ?? CPU);
    }

    public override Tensor forward(Tensor t)
    {
        t = layer1.forward(t);
        t = layer2.forward(t);
        t = layer3.forward(t);
        t = fc.forward(t);
        return LogSoftmax(0).forward(t);
    }
}
```

dayo05 commented:
I found that this shuffler is depend on seed. So, I'll change to other shuffler and enable to use custom shuffler with implementation of IEnumerable.

NiklasGustafsson commented:
@dayo05 -- A couple of comments:

1. This looks really good now. I will be happy to approve this later.
2. There seems to be some problem with the Azure pipelines that are doing the builds for MacOS. We'll have to wait for that to go away before merging.
3. Before we merge, I would like to see at least one of the examples in this repo modified to use this API instead of the version I put together for temporary purposes.

dayo05 commented:
I make fisher yates shuffler as default because I thought that default value is friendly for beginner and previous shuffler is not good for beginner who has less amount of dataset. But, I left previous shuffler and allow to call like this
```cs
var data = new DataLoader(dataset, batchSize, new FastShuffler(dataset.Count), torch.CUDA);
```

dayo05 commented:
https://github.com/dayo05/DataLoaderExample/tree/master
I write example usage for this commit

NiklasGustafsson commented:
> https://github.com/dayo05/DataLoaderExample/tree/master I write example usage for this commit See Fruits360.cs, Program.cs

That's great. Could you take the MNIST example in this repo and convert it to use your API?

dayo05 commented:
@NiklasGustafsson  I updated here https://github.com/dayo05/DataLoaderExample/tree/master
Exist on MNIST.cs, MNISTReader.cs
I edited from torchsharp main repo source

dayo05 commented:
@NiklasGustafsson I fix it

NiklasGustafsson commented:
Okay, I'm going to merge this now. Let's still get some of the examples in this repository using the DataLoader API.
